File: Python/Spiral Matrix II.py
class Solution:
    # @return a list of lists of integer
    def generateMatrix(self, n):
        seq = [[0 for col in range(n)] for row in range(n)]
        up = 0
        left = 0
        down = n - 1
        right = n - 1
        direct = 0
        cnt = 1
        while True:
            if direct == 0:
                for i in range(left, right+1):
                    seq[up][i] = cnt
                    cnt += 1
                up += 1
            if direct == 1:
                for i in range(up, down+1):
                    seq[i][right] = cnt
                    cnt += 1
                right -= 1
            if direct == 2:
                for i in range(right, left-1, -1):
                    seq[down][i] = cnt
                    cnt += 1
                down -= 1
            if direct == 3:
                for i in range(down, up-1, -1):
                    seq[i][left] = cnt
                    cnt += 1
                left += 1
            if up > down or left > right:
                return seq
            direct = (direct+1) % 4

File: C++/Spiral-Matrix-II.cpp
class Solution {
public:
    vector<vector<int> > generateMatrix(int n) {
        vector<vector<int> > matrix(n, vector<int>(n, 0));
        int cur = 0, x = 0, y = 0;
        int left = 0, right = n - 1, up = 0, down = n - 1;
        while (cur < n * n) {
        	while (cur < n * n && y <= right && matrix[x][y] == 0) {
        		matrix[x][y ++] = ++ cur;
        	}
        	y --, up ++, x ++;
        	while (cur < n * n && x <= down && matrix[x][y] == 0) {
        		matrix[x ++][y] = ++ cur;
        	}
        	x --, y --, right --;
        	while (cur < n * n && y >= 0 && matrix[x][y] == 0) {
        		matrix[x][y --] = ++ cur;
        	}
        	y ++, x --, down --;
        	while (cur < n * n && x >= 0 && matrix[x][y] == 0) {
        		matrix[x --][y] = ++ cur;
        	}
        	x ++, y ++, left ++;
        }
        return matrix;
    }
};